core/cog-shell.c

```diff
@@ -716,5 +737,121 @@ CogViewport *
 cog_shell_get_viewport(CogShell *shell)
 {
     g_return_val_if_fail(COG_IS_SHELL(shell), NULL);
-    return PRIV(shell)->viewport;
+
+    gsize        size = cog_shell_get_n_viewports(shell);
+    CogViewport *viewport;
+
+    // Create and add the default viewport if this doesn't exist already
+    if (size == 0) {
+        viewport = cog_viewport_new();
+        cog_shell_add_viewport(shell, viewport);
+    } else
+        viewport = cog_shell_get_nth_viewport(shell, COG_SHELL_DEFAULT_VIEWPORT_INDEX);
+    return viewport;
+}
+
+GPtrArray *
+cog_shell_get_viewports(CogShell *shell)
```

Comment:
I'm not a fan of this, because in general it's a bad idea to use `GPtrArray` in public API (it is not introspectable) but provided that we would do #634 later I think we can land this patch as-is.